diff --git a/libavcodec/vitadec_video.c b/libavcodec/vitadec_video.c
index 8e0b634..48b63f3 100644
--- a/libavcodec/vitadec_video.c
+++ b/libavcodec/vitadec_video.c
@@ -270,7 +270,7 @@ fail:
     return false;
 }

-static void do_mem_free(VitaDecodeBufferFreeParams *p)
+static void __attribute__((optimize("no-optimize-sibling-calls"))) do_mem_free(VitaDecodeBufferFreeParams *p)
 {
     sceKernelFreeMemBlock(*p->mb);
 }
@@ -285,7 +285,7 @@ static bool do_unmap_open(VitaDecodeBufferAllocParams *p)
     return true;
 }

-static void do_unmap_close(VitaDecodeBufferFreeParams *p)
+static void __attribute__((optimize("no-optimize-sibling-calls"))) do_unmap_close(VitaDecodeBufferFreeParams *p)
 {
     sceCodecEngineCloseUnmapMemBlock(*p->mb);
 }
@@ -302,7 +302,7 @@ static bool do_vaddr_alloc(VitaDecodeBufferAllocParams *p)
     return true;
 }

-static void do_vaddr_free(VitaDecodeBufferFreeParams *p)
+static void __attribute__((optimize("no-optimize-sibling-calls"))) do_vaddr_free(VitaDecodeBufferFreeParams *p)
 {
     SceUID *unmap = p->ref;
     SceUIntVAddr *vaddr = p->ptr;
@@ -917,17 +917,11 @@ static int vita_decode(AVCodecContext *avctx, AVFrame *frame, int *got_frame, AV
         to_timestamp_vita(avpkt->pts, time_base,  &au.pts);
         to_timestamp_vita(avpkt->dts, time_base, &au.dts);
 
-        ret = sceAvcdecDecodeAuInternal(&ctx->decoder_ctrl, &au, &ctx->decoder_picture_int);
+        ret = sceAvcdecDecode(&ctx->decoder_ctrl, &au, &ap);
         if (ret < 0) {
             av_log(avctx, AV_LOG_ERROR, "vita_h264 decode: decode au failed 0x%x\n", ret);
             return AVERROR_UNKNOWN;
         }
-
-        ret = sceAvcdecDecodeGetPictureWithWorkPictureInternal(&ctx->decoder_ctrl, &ap, &wap, &ctx->decoder_picture_int);
-        if (ret < 0) {
-            av_log(avctx, AV_LOG_ERROR, "vita_h264 decode: decode work picture failed 0x%x\n", ret);
-            return AVERROR_UNKNOWN;
-        }
     }
 
     if (!ap.numOfOutput)
