diff --git a/configure b/configure
index 6ca37ca..7a3c02e 100755
--- a/configure
+++ b/configure
@@ -5541,6 +5541,10 @@ case $target_os in
         ;;
     minix)
         ;;
+    horizon)
+        network_extralibs="-lnx"
+        enable section_data_rel_ro
+        ;;
     none)
         ;;
     *)
diff --git a/libavfilter/vf_drawtext.c b/libavfilter/vf_drawtext.c
index abe1ca6..0f1d377 100644
--- a/libavfilter/vf_drawtext.c
+++ b/libavfilter/vf_drawtext.c
@@ -37,7 +37,9 @@
 #if HAVE_UNISTD_H
 #include <unistd.h>
 #endif
+#ifndef __SWITCH__
 #include <fenv.h>
+#endif
 
 #if CONFIG_LIBFONTCONFIG
 #include <fontconfig/fontconfig.h>
@@ -1082,14 +1084,17 @@ static int func_eval_expr_int_format(AVFilterContext *ctx, AVBPrint *bp,
             return AVERROR(EINVAL);
         }
     }
-
+#ifndef __SWITCH__
     feclearexcept(FE_ALL_EXCEPT);
+#endif
     intval = res;
+#ifndef __SWITCH__
 #if defined(FE_INVALID) && defined(FE_OVERFLOW) && defined(FE_UNDERFLOW)
     if ((ret = fetestexcept(FE_INVALID|FE_OVERFLOW|FE_UNDERFLOW))) {
         av_log(ctx, AV_LOG_ERROR, "Conversion of floating-point result to int failed. Control register: 0x%08x. Conversion result: %d\n", ret, intval);
         return AVERROR(EINVAL);
     }
+#endif
 #endif
 
     if (argc == 3)
diff --git a/libavformat/os_support.c b/libavformat/os_support.c
index 15cea7f..28714c3 100644
--- a/libavformat/os_support.c
+++ b/libavformat/os_support.c
@@ -47,6 +47,7 @@
 #if !HAVE_INET_ATON
 #include <stdlib.h>
 
+/*
 static int inet_aton(const char *str, struct in_addr *add)
 {
     unsigned int add1 = 0, add2 = 0, add3 = 0, add4 = 0;
@@ -61,6 +62,7 @@ static int inet_aton(const char *str, struct in_addr *add)
 
     return 1;
 }
+*/
 #endif /* !HAVE_INET_ATON */
 
 int ff_getaddrinfo(const char *node, const char *service,
