diff --git a/libavcodec/tx1_h264.c b/libavcodec/tx1_h264.c
index c45deed971..0335e0e6b5 100644
--- a/libavcodec/tx1_h264.c
+++ b/libavcodec/tx1_h264.c
@@ -461,6 +461,7 @@ static int tx1_h264_end_frame(AVCodecContext *avctx) {
     H264Context            *h = avctx->priv_data;
     TX1H264DecodeContext *ctx = avctx->internal->hwaccel_priv_data;
     AVFrame            *frame = ctx->current_frame;
+    if (frame == NULL) return AVERROR_UNKNOWN;
     FrameDecodeData      *fdd = (FrameDecodeData *)frame->private_ref->data;
     TX1Frame              *tf = fdd->hwaccel_priv;
     AVTX1Map       *input_map = (AVTX1Map *)tf->input_map_ref->data;
